main/src/ca/uwaterloo/flix/util/ProgressBar.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2021 Magnus Madsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.util
+
+class ProgressBar {
+  /**
+    * The characters in the spinner.
+    */
+  private val SpinnerChars = Array("|", "/", "-", "\\")
+
+  /**
+    * The sample rate.
+    */
+  private val SampleRate: Int = 100
```

Comment:
What's the reason for doing an update per X ticks rather than per X time?

Reply:
Here is the overall concern: I want this to be fast. It should not slow down the compiler.

Problem is: I have to call flush (which slows it down) and I am invoked from multiple threads (so I should try to be thread safe). This leads to the very awkward code that is there.

Oh, and as for ticks per time, I assume is too expensive to call `nanoTime` so frequently.

Reply:
I am open to better ideas of course.

Reply:
We could combine them, checking time every X ticks and updating if at least Y time has passed. But maybe that's more complex than it's worth.

One thing that would be nice to have is if one particular function takes a long time to process, it would be shown for a long time on screen (matching user's intuition). That's not the case with either of these implementations.

Reply:
I agree, but only in principle. What you propose is to give some useful information (but that would only be useful to us). 

What I propose is to "show the programmer that something is happening".

Reply:
I am also abusing the same trick as npm and many other tools in that I appear to give the idea that the compiler is working fast. You see lots of text flash by. If I only showed "Typer" then it would appear slow. 

In other words, the purpose of this particular exercise is to give the human emotion of "Ah the compiler is running. Great! And its fast. Something is happening!".

Reply:
Airplanes don't have to be pained, but who would want to fly with a plane where the coating is missing or falling of? Its all about appearances.

Reply:
which is why we should probably add some color and more (useless) information.